File: implementations/node/sdk/src/authentication/access-context-bootstrap.ts
import type { AiSdkCredential, AiSdkCredentialProvider } from "./credential.js";
import { createAiSdkError } from "../errors/sdk-error.js";
import { AiSdkException } from "../errors/sdk-exception.js";

const DEFAULT_ACCESS_CONTEXT_HEADER = "X-Access-Context";
const DEFAULT_TIMEOUT_MS = 30_000;

export interface AiSdkAccessContextBootstrapOptions {
  /** Protected HTTP endpoint that creates one access-context handle. */
  readonly endpoint: URL;
  /** Credential provider used only for this explicit bootstrap request. */
  readonly credentialProvider?: AiSdkCredentialProvider;
  /** Response header carrying the created access-context handle. */
  readonly accessContextHeaderName?: string;
  /** Maximum duration of this one non-retried state-changing request. */
  readonly timeoutMs?: number;
  /** Optional fetch implementation for host integration and deterministic tests. */
  readonly fetch?: typeof fetch;
}

export interface AiSdkAccessContextBootstrapResult {
  readonly accessContext: string;
  readonly headerName: string;
}

/**
 * Creates the first runtime access-context handle over the authenticated public HTTP boundary.
 * This state-changing operation is deliberately never retried automatically.
 */
export class AiSdkAccessContextBootstrapper {
  public static async create(
    options: AiSdkAccessContextBootstrapOptions,
    signal?: AbortSignal,
  ): Promise<AiSdkAccessContextBootstrapResult> {
    const validated = validateOptions(options);
    signal?.throwIfAborted();

    const credential = await getCredential(validated.credentialProvider, signal);
    const headers = new Headers();

    if (credential !== null) {
      headers.set("Authorization", formatAuthorization(credential));
    }

    const controller = new AbortController();
    const onAbort = (): void => controller.abort(signal?.reason);
    signal?.addEventListener("abort", onAbort, { once: true });
    const timeout = setTimeout(
      () => controller.abort(new DOMException("Access-context bootstrap timed out.", "TimeoutError")),
      validated.timeoutMs,
    );

    try {
      let response: Response;
      try {
        response = await validated.fetch(validated.endpoint, {
          method: "POST",
          headers,
          signal: controller.signal,
        });
      } catch (error) {
        if (signal?.aborted === true) {
          signal.throwIfAborted();
        }

        if (controller.signal.aborted) {
          throw sdkException(
            "transport",
            "access_context_bootstrap_timeout",
            "The access-context bootstrap request timed out.",
            error,
          );
        }

        throw sdkException(
          "transport",
          "access_context_bootstrap_transport_failure",
          errorMessage(error),
          error,
        );
      }

      if (response.status === 401) {
        throw sdkException(
          "authentication",
          "access_context_bootstrap_unauthenticated",
          "The access-context bootstrap request was not authenticated.",
        );
      }

      if (response.status === 403) {
        throw sdkException(
          "authorization",
          "access_context_bootstrap_forbidden",
          "The authenticated identity is not allowed to create an access context.",
        );
      }

      if (!response.ok) {
        throw sdkException(
          "remote_failure",
          "access_context_bootstrap_failed",
          `The access-context bootstrap endpoint returned HTTP ${response.status}${response.statusText ? ` (${response.statusText})` : ""}.`,
        );
      }

      const accessContext = response.headers.get(validated.headerName)?.trim();
      if (!isUnambiguousHeaderValue(accessContext)) {
        throw sdkException(
          "invalid_response",
          "access_context_bootstrap_missing_handle",
          `The access-context bootstrap response did not contain one unambiguous '${validated.headerName}' header.`,
        );
      }

      return {
        accessContext,
        headerName: validated.headerName,
      };
    } finally {
      clearTimeout(timeout);
      signal?.removeEventListener("abort", onAbort);
    }
  }
}

type ValidatedOptions = Readonly<{
  endpoint: URL;
  credentialProvider?: AiSdkCredentialProvider;
  headerName: string;
  timeoutMs: number;
  fetch: typeof fetch;
}>;

function validateOptions(options: AiSdkAccessContextBootstrapOptions): ValidatedOptions {
  if (!(options.endpoint instanceof URL) || !["http:", "https:"].includes(options.endpoint.protocol)) {
    throw new TypeError("The access-context bootstrap endpoint must be an absolute HTTP or HTTPS URL.");
  }

  const headerName = (options.accessContextHeaderName ?? DEFAULT_ACCESS_CONTEXT_HEADER).trim();
  if (!isValidHeaderName(headerName)) {
    throw new TypeError("accessContextHeaderName must be a valid non-empty HTTP header name.");
  }

  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
    throw new RangeError("timeoutMs must be a positive finite number.");
  }

  return {
    endpoint: new URL(options.endpoint.toString()),
    ...(options.credentialProvider === undefined
      ? {}
      : { credentialProvider: options.credentialProvider }),
    headerName,
    timeoutMs,
    fetch: options.fetch ?? globalThis.fetch.bind(globalThis),
  };
}

async function getCredential(
  provider: AiSdkCredentialProvider | undefined,
  signal?: AbortSignal,
): Promise<AiSdkCredential | null> {
  if (provider === undefined) {
    return null;
  }

  try {
    const credential = await provider.getCredential(signal);
    signal?.throwIfAborted();
    return credential;
  } catch (error) {
    if (signal?.aborted === true) {
      signal.throwIfAborted();
    }

    throw sdkException(
      "authentication",
      "credential_provider_failure",
      "The SDK credential provider failed to supply a bootstrap credential.",
      error,
    );
  }
}

function formatAuthorization(credential: AiSdkCredential): string {
  const scheme = credential.scheme.trim();
  const value = credential.value.trim();

  if (
    scheme.length === 0 ||
    value.length === 0 ||
    /[\s:\u0000-\u001f\u007f]/u.test(scheme) ||
    /[\r\n]/u.test(value)
  ) {
    throw sdkException(
      "authentication",
      "invalid_credential",
      "The SDK credential provider returned an invalid bootstrap credential.",
    );
  }

  return `${scheme} ${value}`;
}

function isValidHeaderName(value: string): boolean {
  return value.length > 0 && !/[:\r\n\u0000-\u001f\u007f]/u.test(value);
}

function isUnambiguousHeaderValue(value: string | null | undefined): value is string {
  return value !== null &&
    value !== undefined &&
    value.length > 0 &&
    !/[\r\n,]/u.test(value);
}

function sdkException(
  kind: Parameters<typeof createAiSdkError>[0],
  code: string,
  message: string,
  cause?: unknown,
): AiSdkException {
  const details = cause === undefined ? {} : { cause: errorMessage(cause) };
  return new AiSdkException(createAiSdkError(kind, code, message, false, details),
    cause instanceof Error ? { cause } : undefined);
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: implementations/python/sdk/src/multiplexed_ai_sdk/access_context_bootstrap.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Mapping
from urllib.parse import urlparse

from .authentication import AiSdkCredential, AiSdkCredentialProvider
from .errors import AiSdkError, AiSdkException

_DEFAULT_ACCESS_CONTEXT_HEADER = "X-Access-Context"
_DEFAULT_TIMEOUT_SECONDS = 30.0


@dataclass(frozen=True)
class AiSdkAccessContextBootstrapOptions:
    """Configuration for one explicit access-context bootstrap request."""

    endpoint: str
    credential_provider: AiSdkCredentialProvider | None = None
    access_context_header_name: str = _DEFAULT_ACCESS_CONTEXT_HEADER
    timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS

    def __post_init__(self) -> None:
        parsed = urlparse(self.endpoint)
        if parsed.scheme not in {"http", "https"} or not parsed.netloc:
            raise ValueError(
                "The access-context bootstrap endpoint must be an absolute HTTP or HTTPS URL."
            )

        header_name = self.access_context_header_name.strip()
        if (
            not header_name
            or ":" in header_name
            or "\r" in header_name
            or "\n" in header_name
            or any(ord(character) < 32 or ord(character) == 127 for character in header_name)
        ):
            raise ValueError(
                "access_context_header_name must be a valid non-empty HTTP header name."
            )

        if self.timeout_seconds <= 0:
            raise ValueError("timeout_seconds must be greater than zero")


@dataclass(frozen=True)
class AiSdkAccessContextBootstrapResult:
    access_context: str
    header_name: str


class AiSdkAccessContextBootstrapper:
    """Creates the initial runtime access-context handle over authenticated HTTP.

    The POST is deliberately explicit and is never retried automatically because
    access-context creation changes server state. Subsequent handle rotation is owned
    by the physical SDK transport.
    """

    @staticmethod
    async def create(
        options: AiSdkAccessContextBootstrapOptions,
    ) -> AiSdkAccessContextBootstrapResult:
        if options is None:
            raise ValueError("options is required")

        credential = await _get_credential(options.credential_provider)
        headers: dict[str, str] = {}
        if credential is not None:
            headers["Authorization"] = _format_authorization(credential)

        try:
            async with _create_http_client(options.timeout_seconds) as client:
                response = await client.post(options.endpoint, headers=headers or None)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            if _is_timeout(exc):
                raise _sdk_exception(
                    "transport",
                    "access_context_bootstrap_timeout",
                    "The access-context bootstrap request timed out.",
                    exc,
                ) from exc
            raise _sdk_exception(
                "transport",
                "access_context_bootstrap_transport_failure",
                str(exc) or type(exc).__name__,
                exc,
            ) from exc

        status_code = int(getattr(response, "status_code", 0))
        if status_code == 401:
            raise _sdk_exception(
                "authentication",
                "access_context_bootstrap_unauthenticated",
                "The access-context bootstrap request was not authenticated.",
            )
        if status_code == 403:
            raise _sdk_exception(
                "authorization",
                "access_context_bootstrap_forbidden",
                "The authenticated identity is not allowed to create an access context.",
            )
        if status_code < 200 or status_code >= 300:
            reason = getattr(response, "reason_phrase", "")
            suffix = f" ({reason})" if isinstance(reason, str) and reason.strip() else ""
            raise _sdk_exception(
                "remote_failure",
                "access_context_bootstrap_failed",
                f"The access-context bootstrap endpoint returned HTTP {status_code}{suffix}.",
            )

        header_name = options.access_context_header_name.strip()
        values = _header_values(getattr(response, "headers", {}), header_name)
        normalized = tuple(dict.fromkeys(value.strip() for value in values if _valid_header_value(value)))

        if len(normalized) != 1:
            raise _sdk_exception(
                "invalid_response",
                "access_context_bootstrap_missing_handle",
                f"The access-context bootstrap response did not contain one unambiguous '{header_name}' header.",
            )

        return AiSdkAccessContextBootstrapResult(
            access_context=normalized[0],
            header_name=header_name,
        )


async def _get_credential(
    provider: AiSdkCredentialProvider | None,
) -> AiSdkCredential | None:
    if provider is None:
        return None

    try:
        return await provider.get_credential()
    except asyncio.CancelledError:
        raise
    except Exception as exc:
        raise _sdk_exception(
            "authentication",
            "credential_provider_failure",
            "The SDK credential provider failed to supply a bootstrap credential.",
            exc,
        ) from exc


def _format_authorization(credential: AiSdkCredential) -> str:
    scheme = credential.scheme.strip()
    value = credential.value.strip()

    if (
        not scheme
        or not value
        or any(character.isspace() or character == ":" or ord(character) < 32 or ord(character) == 127 for character in scheme)
        or "\r" in value
        or "\n" in value
    ):
        raise _sdk_exception(
            "authentication",
            "invalid_credential",
            "The SDK credential provider returned an invalid bootstrap credential.",
        )

    return f"{scheme} {value}"


def _header_values(headers: Mapping[str, str] | Any, name: str) -> tuple[str, ...]:
    get_list = getattr(headers, "get_list", None)
    if callable(get_list):
        raw = get_list(name)
        if isinstance(raw, list):
            return tuple(value for value in raw if isinstance(value, str))

    expected = name.lower()
    values = [
        value
        for header_name, value in getattr(headers, "items", lambda: ())()
        if isinstance(header_name, str)
        and header_name.lower() == expected
        and isinstance(value, str)
    ]
    return tuple(values)


def _valid_header_value(value: str | None) -> bool:
    return (
        isinstance(value, str)
        and bool(value.strip())
        and "\r" not in value
        and "\n" not in value
    )


def _is_timeout(exc: Exception) -> bool:
    return isinstance(exc, TimeoutError) or "Timeout" in type(exc).__name__


def _create_http_client(timeout_seconds: float) -> Any:
    import httpx2

    return httpx2.AsyncClient(timeout=timeout_seconds)


def _sdk_exception(
    kind: str,
    code: str,
    message: str,
    cause: Exception | None = None,
) -> AiSdkException:
    details = {} if cause is None else {"cause": str(cause) or type(cause).__name__}
    return AiSdkException(
        AiSdkError(
            kind=kind,  # type: ignore[arg-type]
            code=code,
            message=message,
            retryable=False,
            details=details,
        )
    )
